pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,420 @@
+		result = append(result, "sha256:"+hex.EncodeToString(keyHash[:]))
+
+		result = append(result, keyObj.Subjects()...)
+	}
+
+	if v.DSSEObj.PayloadHash != nil {
+		payloadHashKey := strings.ToLower(fmt.Sprintf("%s:%s", *v.DSSEObj.PayloadHash.Algorithm, *v.DSSEObj.PayloadHash.Value))
+		result = append(result, payloadHashKey)
+	}
+
+	if v.DSSEObj.EnvelopeHash != nil {
+		envelopeHashKey := strings.ToLower(fmt.Sprintf("%s:%s", *v.DSSEObj.EnvelopeHash.Algorithm, *v.DSSEObj.EnvelopeHash.Value))
+		result = append(result, envelopeHashKey)
+	}
+
+	if v.env == nil {
+		log.Logger.Info("DSSEObj content or DSSE envelope is nil, returning partial set of keys")
+		return result, nil
+	}
+
+	switch v.env.PayloadType {
+	case in_toto.PayloadType:
+
+		if v.env.Payload == "" {
+			log.Logger.Info("DSSEObj DSSE payload is empty")
+			return result, nil
+		}
+		decodedPayload, err := v.env.DecodeB64Payload()
+		if err != nil {
+			return result, fmt.Errorf("could not decode envelope payload: %w", err)
+		}
+		statement, err := parseStatement(decodedPayload)
+		if err != nil {
+			return result, err
+		}
+		for _, s := range statement.Subject {
+			for alg, ds := range s.Digest {
+				result = append(result, alg+":"+ds)
+			}
+		}
+		// Not all in-toto statements will contain a SLSA provenance predicate.
+		// See https://github.com/in-toto/attestation/blob/main/spec/README.md#predicate
+		// for other predicates.
+		if predicate, err := parseSlsaPredicate(decodedPayload); err == nil {
+			if predicate.Predicate.Materials != nil {
+				for _, s := range predicate.Predicate.Materials {
+					for alg, ds := range s.Digest {
+						result = append(result, alg+":"+ds)
+					}
+				}
+			}
+		}
+	default:
+		log.Logger.Infof("Unknown DSSE envelope payloadType: %s", v.env.PayloadType)
+	}
+	return result, nil
+}
+
+func parseStatement(p []byte) (*in_toto.Statement, error) {
+	ps := in_toto.Statement{}
+	if err := json.Unmarshal(p, &ps); err != nil {
+		return nil, err
+	}
+	return &ps, nil
+}
+
+func parseSlsaPredicate(p []byte) (*in_toto.ProvenanceStatement, error) {
+	predicate := in_toto.ProvenanceStatement{}
+	if err := json.Unmarshal(p, &predicate); err != nil {
+		return nil, err
+	}
+	return &predicate, nil
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	it, ok := pe.(*models.DSSE)
+	if !ok {
+		return errors.New("cannot unmarshal non DSSE v0.0.1 type")
+	}
+
+	dsseObj := &models.DSSEV001Schema{}
+
+	if err := types.DecodeEntry(it.Spec, dsseObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := dsseObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	// either we have just proposed content or the canonicalized fields
+	if dsseObj.ProposedContent == nil {
+		// then we need canonicalized fields, and all must be present (if present, they would have been validated in the above call to Validate())
+		if dsseObj.EnvelopeHash == nil || dsseObj.PayloadHash == nil || len(dsseObj.Signatures) == 0 {
```

Comment:
To confirm, because we have Insertable(), we know that the DSSEObj must have valid signatures at this point, correct? As in, it's not possible to insert a DSSE envelope with invalid signatures because Insertable checks that proposed content is set, and then we check the signatures if proposed content is set

Reply:
Insertable is called *after* Unmarshal on the insert path.